sap-service-operator/src/main/java/com/sap/cloud/environment/servicebinding/FileSystemWatcherCache.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved.
+ */
+
+package com.sap.cloud.environment.servicebinding;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.sap.cloud.environment.servicebinding.api.ServiceBinding;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+
+class FileSystemWatcherCache implements DirectoryBasedCache
+{
+    @Nonnull
+    private static final Collection<WatchEvent.Kind<?>> MODIFICATION_EVENTS =
+        Arrays.asList(ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
+
+    @Nonnull
+    private final Function<Path, ServiceBinding> serviceBindingLoader;
+    @Nonnull
+    // internal for simplified testing
+    final Map<Path, ServiceBinding> cachedServiceBindings = new HashMap<>();
+    @Nonnull
+    // internal for simplified testing
+    final Map<Path, WatchKey> directoryWatchKeys = new HashMap<>();
+    @Nonnull
+    private final WatchService watchService;
+
+    public FileSystemWatcherCache( @Nonnull final Function<Path, ServiceBinding> serviceBindingLoader )
+    {
+        this.serviceBindingLoader = serviceBindingLoader;
+
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch( final IOException e ) {
+            throw new IllegalStateException(
+                String.format("Unable to create new instance of '%s'.", WatchService.class.getSimpleName()),
+                e);
+        }
+    }
+
+    @Nonnull
+    @Override
+    public synchronized List<ServiceBinding> getServiceBindings( @Nonnull final Collection<Path> directories )
+    {
+        removeOutdatedWatchKeys(directories);
+        removeOutdatedServiceBindings(directories);
+        return directories
+            .stream()
+            .peek(this::renewCachedServiceBindingIfNeeded)
+            .map(cachedServiceBindings::get)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+    }
+
+    private void removeOutdatedWatchKeys( @Nonnull final Collection<Path> directoriesOfInterest )
+    {
+        for( final Map.Entry<Path, WatchKey> entry : directoryWatchKeys.entrySet() ) {
+            if( directoriesOfInterest.contains(entry.getKey()) ) {
+                continue;
+            }
+
+            entry.getValue().cancel();
+            directoryWatchKeys.remove(entry.getKey());
+        }
```

Comment:
(Minor)
Please consider Iterator API usage for simultaneous iterating and removing.

Reply:
Not sure if I like the two proposed alternatives better than the current code... 🤔

```java
final Iterator<Map.Entry<Path, WatchKey>> it = directoryWatchKeys.entrySet().iterator();
while (it.hasNext()) {
    final Map.Entry<Path, WatchKey> entry = it.next();
    if (directoriesOfInterest.contains(entry.getKey())) {
        continue;
    }

    entry.getValue().cancel();
    it.remove();
}
```

vs. 

```java
directoryWatchKeys.entrySet().removeIf(entry -> {
    if (!directoriesOfInterest.contains(entry.getKey())) {
        return false;
    }

    entry.getValue().cancel();
    return true;
});
```

vs.

```java
for( final Map.Entry<Path, WatchKey> entry : directoryWatchKeys.entrySet() ) {
    if( directoriesOfInterest.contains(entry.getKey()) ) {
        continue;
    }

    entry.getValue().cancel();
    directoryWatchKeys.remove(entry.getKey());
}
```

Reply:
I asked, because I would actually expect a runtime exception for code in the form of...
```java
for( Map#entrySet() ) {
  Map#remove(...);
}
```

Have you checked that an exception (can) never occur?

Reply:
I first had a variant that was using the stream API, which indeed caused an exception.
The current implementation, however, is working as tested [here](https://github.com/SAP/btp-environment-variable-access/blob/feat/file-system-aware-cache/sap-service-operator/src/test/java/com/sap/cloud/environment/servicebinding/FileSystemWatcherCacheTest.java#L194-L215).